Redis receiver

pmcollins opened the pull request:
**Description:** Add Redis receiver.

Commit:
Work in Progress: Redis receiver

Todo:
1) Error handling: currently bailing out on any error. Need to add resiliency.
2) Improved organization: everything is in one sprawling package.
3) Verification of Redis metric types.
4) Licensing headers.
5) Comments.
6) Logging.

Commit:
Work in Progress: Redis receiver

Todo:
1) Error handling: currently bailing out on any error. Need to add resiliency.
2) Improved organization: everything is in one sprawling package.
3) Verification of Redis metric types.
4) Licensing headers.
5) Comments.
6) Logging.

Commit:
wip: redis receiver: fix typo, use slice literal

Commit:
wip: redis receiver: fix typo, use slice literal

Commit:
redis receiver: add logging, error handling, thread safety

Commit:
redis receiver: add logging, error handling, thread safety

Commit:
redis receiver: rename stragging 'tickable' variables to runnable

Commit:
redis receiver: rename stragging 'tickable' variables to runnable

pmcollins commented on receiver/redisreceiver/metric_functions.go:
```diff
@@ -0,0 +1,231 @@
+package redisreceiver
+
+func uptimeInSeconds() *redisMetric {
```
It has been suggested that I turn this data into a map. Looking into it.

pmcollins commented on receiver/redisreceiver/metrics_service_test.go:
```diff
@@ -0,0 +1,70 @@
+	require.Nil(t, err)
+	require.Equal(t, 29, len(md.Metrics))
+}
+
```
Need to test error/failure handling as well.

pmcollins commented on receiver/redisreceiver/receiver.go:
```diff
@@ -0,0 +1,71 @@
+}
+
+func (r redisReceiver) Start(host component.Host) error {
+	r.Lock()
```
I modeled this on the other receivers but would like to learn more about why the mutex and startOnce are needed.

tigrannajaryan replied:
It should not be needed. `Start` is called once only.

@pjanotti any thoughts on this? I saw the same pattern in `signalfxreceiver`, so perhaps there is something I am missing.

(We do have "once" protection in opencensusreceiver, but there it is needed because it implements trace and metrics receiver via single instance and internal `start` function may be called twice. Everywhere else I do not think we need such protection).

pmcollins replied:
Ok, will remove for now. Thanks.

flands commented on receiver/redisreceiver/config.go:
```diff
@@ -0,0 +1,8 @@
+import "github.com/open-telemetry/opentelemetry-collector/config/configmodels"
+
+type Config struct {
+	configmodels.ReceiverSettings `mapstructure:",squash"`
```
Please add a README with configuration information -- see other receivers for an example

pmcollins replied:
Sounds good. Will do.

tigrannajaryan commented on receiver/redisreceiver/converters.go:
```diff
@@ -0,0 +1,29 @@
+import (
+	"strconv"
+
+	metricsProto "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
```
We usually alias this import to `metricspb` elsewhere in the codebase.
```suggestion
	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
```

pmcollins replied:
Fixed.

tigrannajaryan commented on receiver/redisreceiver/api_parser.go:
```diff
@@ -0,0 +1,39 @@
+
+import "strings"
+
+type apiParser struct {
```
Please add comments to explain the purpose of non-trivial structs and functions.

pmcollins replied:
Done.

tigrannajaryan commented on receiver/redisreceiver/metrics_service.go:
```diff
@@ -0,0 +1,37 @@
+package redisreceiver
+
+import (
+	resourceProto "github.com/census-instrumentation/opencensus-proto/gen-go/resource/v1"
```
For uniformity with the rest of the codebase:
```suggestion
	resourcepb "github.com/census-instrumentation/opencensus-proto/gen-go/resource/v1"
```

pmcollins replied:
Fixed.

tigrannajaryan commented on receiver/redisreceiver/metric_functions.go:
```diff
@@ -0,0 +1,231 @@
+func uptimeInSeconds() *redisMetric {
+	return &redisMetric{
+		key:        "uptime_in_seconds",
+		units:      "seconds",
```
Since this eventually gets assigned to `MetricDescriptor.Unit` it needs to follow conventions defined in http://unitsofmeasure.org/ucum.html

@bogdandrutu do I understand it correct that we should use "c/s" form for unit, i.e. for seconds we should use "s". Is that right?

pmcollins replied:
I kind of punted on the units for this draft, but will make sure they are consistent with the spec.

pmcollins replied:
Fixed.

tigrannajaryan commented on receiver/redisreceiver/redis_metrics.go:
```diff
@@ -0,0 +1,157 @@
+	gaugeDistribution
+	cumulativeInt
+	cumulativeDouble
+)
```
Is there any reason that we redeclare this instead of using MetricDescriptor_Type type? The code elsewhere assumes they must be equivalent and uses just type-cast for converting between metricType and MetricDescriptor_Type. In that case what do we gain by having this separate declaration?

pmcollins replied:
I wanted to keep this independent of protobuf because it was my understanding that there would be a new data model coming soon that was encoding agnostic, so I didn't want to potentially have to change these values in a bunch of places. Happy to use MetricDescriptor_Type directly though if you think that would be better at this time.

tigrannajaryan replied:
Thanks, I understand now the intent. However, because we simply use type conversion between metricType and MetricDescriptor_Type they are inherently tightly coupled although it may not appear so when glancing at the code. If the intent is to make redis metric declarations encoding agnostic then it would be better to have conversion function from metricType to MetricDescriptor_Type and avoid using type conversion.
That is likely an overkill though. We are likely fine with relying on internal data types (and we can assume MetricDescriptor_Type is an internal data type - although the assumption is temporary) and just use MetricDescriptor_Type directly. Once we have the new internal data types we can just update the declaration here. When that happens the compiler will tell us if there are any type mismatches elsewhere, but only if we avoid using type conversions.

Anyway, this is a minor nit for now. Eventually I believe we should feel free to use internal data types within our codebase.

pmcollins replied:
Sounds good. Replaced with MetricDescriptor_Type.

tigrannajaryan commented on receiver/redisreceiver/redis_metrics.go:
```diff
@@ -0,0 +1,157 @@
+func buildSingleProtoMetric(strVal string, redisMetric *redisMetric) (*metricsProto.Metric, error) {
+	f := getPointConverter(redisMetric)
+	if f == nil {
+		return nil, errors.New("oops") // todo constant?
```
Presumably this can never happen because we don't have redisMetrics of any type that is not an int or double. Can we somehow ensure that if we do we can a compile error or we have a test that ensures this is not possible? This would prevent future problems if someone adds a redisMetric with a different type to getDefaultRedisMetrics.

tigrannajaryan reviewed (commented):
@pmcollins Nice work. I left a few comments.

A general concern about the code is lack of explanatory comments. Would be great to add some.

pmcollins commented:
> @pmcollins Nice work. I left a few comments.
> 
> A general concern about the code is lack of explanatory comments. Would be great to add some.

Thanks for the review. Will add comments.

Commit:
Update Redis receiver based on prelim feedback

Commit:
Update Redis receiver based on prelim feedback

Commit:
Add comments to Redis receiver

Commit:
Merge branch 'redis' of https://github.com/signalfx/opentelemetry-collector-contrib into redis

jrcamp commented on receiver/redisreceiver/README.md:
```diff
@@ -0,0 +1,40 @@
+# Redis Receiver
```
@tigrannajaryan There might be config options or metric names/labels that we want to change as we learn more when implementing additional monitors/receivers that would break compatibility. Should we mark this as beta/experimental somehow so we don't have to worry about making breaking changes yet?

Commit:
Run go fmt on redis receiver

Commit:
Add copyright to redisreceiver files

Commit:
Go fmt redisreceiver

Commit:
Fix import groupings for make ci

pmcollins commented:
I signed it

Commit:
Refactor and cleanup lint issues

pmcollins committed:
Merge branch 'master' into redis

Commit:
Redis receiver otelcol version bump

Commit:
Redis receiver tweaks

pjanotti commented on receiver/redisreceiver/info.go:
```diff
@@ -0,0 +1,78 @@
+			warnings = append(warnings, fmt.Errorf("info key not found: %v", redisMetric.key))
+			continue
+		}
+		if len(strVal) == 0 {
```
nit: this repo in general compares against ""

pmcollins replied:
Noted, fixed.

Commit:
Update receiver/redisreceiver/README.md

Co-Authored-By: Paulo Janotti <pjanotti@splunk.com>

pmcollins commented on receiver/redisreceiver/client.go:
```diff
@@ -0,0 +1,52 @@
+	"github.com/go-redis/redis/v7"
+)
+
+// Interface for a Redis client. Implementation can be fake or real.
```
Exactly. I have added a comment to clarify.

pmcollins commented on receiver/redisreceiver/keyspace.go:
```diff
@@ -0,0 +1,53 @@
+	for _, pairStr := range pairs {
+		pair := strings.Split(pairStr, "=")
+		key := pair[0]
+		val, err := strconv.Atoi(pair[1])
```
Good point, fixed.

pmcollins commented on receiver/redisreceiver/redis_runnable.go:
```diff
@@ -0,0 +1,84 @@
+	}
+
+	md := newMetricsData(metrics)
+	return r.metricsConsumer.ConsumeMetricsData(r.ctx, *md)
```
Thanks, done.

pmcollins commented on receiver/redisreceiver/redis_metric.go:
```diff
@@ -0,0 +1,66 @@
+func (m *redisMetric) parsePoint(strVal string) (*metricspb.Point, error) {
+	switch m.mdType {
+	case metricspb.MetricDescriptor_CUMULATIVE_INT64, metricspb.MetricDescriptor_GAUGE_INT64:
+		return strToInt64Point(strVal)
```
Agree and fixed.

Commit:
Redis receiver: Expand README, add obsreport metrics.

Commit:
Tweak Redis runnable test formatting

pmcollins committed:
Merge branch 'master' into redis

Commit:
Update redis collector to use latest otelcol api

pmcollins committed:
Merge branch 'master' into redis

Commit:
Change redis refresh interval to collection interval

tigrannajaryan commented:
Please use imperative mood in the subject line of commit message first line and add more details to both commit message and PR description to explain what the change does.

If you didn't yet have a chance please read:
- The general contributing guidelines at https://github.com/open-telemetry/community/blob/master/CONTRIBUTING.md
- Collector contributing guidelines at https://github.com/open-telemetry/opentelemetry-collector/blob/master/CONTRIBUTING.md
- How to write good commit messages at https://chris.beams.io/posts/git-commit/

tigrannajaryan commented on receiver/redisreceiver/README.md:
```diff
@@ -0,0 +1,89 @@
+instance, build metrics from that data, and send them to the next consumer at a
+configurable interval.
+
+Status: beta
```
👍

tigrannajaryan commented on receiver/redisreceiver/redis_runnable.go:
```diff
@@ -0,0 +1,105 @@
+	return nil
+}
+
+// Runs intermittently, querying Redis and building Metrics to send to the
```
>Runs intermittently

This is a bit confusing. This function performs a single run, single cycle of collection from Redis, right? When I read the comment I get the impression that Run() function internally performs multiple run cycles.

Perhaps reword a bit, like this:
>// Run is called periodically, ...

pmcollins replied:
Better, fixed.

tigrannajaryan commented on receiver/redisreceiver/redis_svc.go:
```diff
@@ -0,0 +1,54 @@
+			continue
+		}
+		pair := strings.Split(line, ":")
+		if len(pair) == 2 { // defensive, should always == 2
```
👍

tigrannajaryan commented on receiver/redisreceiver/README.md:
```diff
@@ -0,0 +1,89 @@
+
+### Overview
+
+The Redis receiver is designed to retrieve Redis INFO data from a single Redis
+instance, build metrics from that data, and send them to the next consumer at a
+configurable interval.
```
Nice overview.

tigrannajaryan commented on receiver/redisreceiver/README.md:
```diff
@@ -0,0 +1,89 @@
+
+### Details
+
+The Redis INFO command returns information and statistics about a Redis
+server (see [https://redis.io/commands/info](https://redis.io/commands/info) for
+details).
+The Redis receiver extracts values from the result and converts them to open
+telemetry metrics. Details about the metrics produced by the Redis receiver
+can be found by browsing [metric_functions.go](metric_functions.go).
+
+For example, one of the fields returned by the Redis INFO command is
+`used_cpu_sys` which indicates the system CPU consumed by the Redis server,
+expressed in seconds, since the start of the Redis instance.
```
Nit: please use consistent line spacing for paragraphs (the first 2 paragraphs do not have an empty line between).

pmcollins replied:
Done (it's just one paragraph).

tigrannajaryan commented on receiver/redisreceiver/README.md:
```diff
@@ -0,0 +1,89 @@
+
+_Required._
+
+### refresh_interval
```
Is it collection_interval or refresh_interval?

pmcollins replied:
Should be collection interval -- I renamed the field and missed this file. Fixed.

pmcollins commented on receiver/redisreceiver/proto.go:
```diff
@@ -0,0 +1,136 @@
+}
+
+// TODO: Maybe this should be moved to a general purpose utility if we don't have one already
+func timeToTimestamp(t *time.Time) *timestamp.Timestamp {
```
Nice, fixed.

pmcollins commented on receiver/redisreceiver/proto.go:
```diff
@@ -0,0 +1,136 @@
+
+	if redisMetric.mdType == metricspb.MetricDescriptor_CUMULATIVE_INT64 ||
+		redisMetric.mdType == metricspb.MetricDescriptor_CUMULATIVE_DOUBLE {
+		startTime = &timestamp.Timestamp{} // todo: not sure about this
```
Nice, thanks. I have made a change to track the server start time as you suggest.

Commit:
Add server start time to cumulative metrics

Cumulative metrics have an implied start time. This change keeps track
of the server start time so it can be applied to cumulative metrics.

Also fixes issues raised in PR feedback.

pmcollins committed:
Merge branch 'master' into redis

Commit:
Fix redis receiver issues from code review

* Use named return values for errors that should be treated as warnings
* Handle malformed keyspace key value pairs
* Update README to use env var in example config
* Remove unnecessary EndMetricsReceiveOp in redis runnable

pmcollins committed:
Merge branch 'master' into redis

Commit:
Update go sum for redis

Commit:
Update otelcol version for Redis receiver

pjanotti reviewed (approved):
LGTM. Thanks @pmcollins!
